agent/taskresource/credentialspec/credentialspec_windows.go

```diff
@@ -455,26 +466,42 @@ func (cs *CredentialSpecResource) writeS3File(writeFunc func(file oswrapper.File
 	if err != nil {
 		return err
 	}
-	defer temp.Close()
 
 	err = writeFunc(temp)
 	if err != nil {
 		return err
 	}
 
-	err = temp.Chmod(os.FileMode(filePerm))
+	// Persist the file to disk.
+	err = temp.Sync()
 	if err != nil {
 		return err
 	}
 
-	// Persist the file to disk.
-	err = temp.Sync()
+	input, err := ioutil.ReadFile(temp.Name())
+	if err != nil {
+		return err
+	}
+
+	err = ioutil.WriteFile(filePath, input, 0644)
+	if err != nil {
+		return err
+	}
+
+	err = acl.Chmod(filePath, filePerm);
+	if err != nil {
+		return err
+	}
+
+	err = temp.Close()
 	if err != nil {
+		seelog.Errorf("Error while closing the handle to file:%v", temp.Name())
 		return err
 	}
 
-	err = cs.os.Rename(temp.Name(), filePath)
+	err = os.Remove(temp.Name())
```

Comment:
Looking at this, it seems like the order in which the chmod / rename happens is the important factor here.

If you move the chmod after the old Rename function, wouldn't that accomplish the same thing? Or is there something else happening here?

Reply:
The problem is not with file permissions. It was failing because some other process is still accessing the memory .  I am guessing it might be S3 library . I have seen similar issues in the past . Moving chmod after rename will not fix the issue.

Reply:
I see two changes here: (1) chmod now happens after the file operation (2) the file operation is changed from os.Rename() to explicitly reading the file into memory and writing it out.

To me, (1) makes the most since for what is solving the issue. I don't understand how memory  access, other processes, or S3's library come into play here, could you explain?

Reply:
>  I have seen similar issues in the past 

Alternatively, is there a best practice for this situation if its common to windows? I don't have confidence that we understand the underlying issue here.

Recall that agent is the only thing that has touched the file at this point in time, so if something is mismanaging the file handle, its our problem.

Reply:
same here, if you are going to use os, plz use the one passed into the credentialspec object